Add Password Strength Checker to Registration

stevenelleman opened the pull request.

stevenelleman commented:
<img width="429" alt="Screenshot 2024-11-03 at 9 06 16 PM" src="https://github.com/user-attachments/assets/243d6c2c-e648-40c1-bf17-40322357cda7">

<img width="432" alt="Screenshot 2024-11-03 at 9 06 03 PM" src="https://github.com/user-attachments/assets/04fa5967-da3c-46b7-be7b-4c6835769121">

stevenelleman commented:
Added more explicit first warning and cleaned up a bit: 
<img width="443" alt="image" src="https://github.com/user-attachments/assets/9682d643-5fd7-4a66-ace5-03442a83a32c">

AndrewCLu commented on apps/frontend/package.json:
```diff
@@ -48,6 +48,7 @@
     "tailwindcss": "^3.4.1",
-    "typescript": "^5"
+    "typescript": "^5",
+    "zxcvbn": "^4.4.2"
```
does it make sense to use typescript version of this instead? https://www.npmjs.com/package/@zxcvbn-ts/core

stevenelleman replied:
Thanks for finding, looked and didn't find it originally.

AndrewCLu commented on apps/frontend/src/pages/register.tsx:
```diff
@@ -166,6 +169,22 @@ const Register: React.FC = () => {
+
+    // Check password strength
+    const passwordCheck = zxcvbn(password);
+    if (passwordCheck && passwordCheck.score < 3) {
```
why 3? seems arbitrary, can you add a comment and link to some docs explaining this?

AndrewCLu commented on apps/frontend/src/pages/register.tsx:
```diff
@@ -166,6 +169,22 @@ const Register: React.FC = () => {
+      if (passwordCheck.feedback?.warning) {
+        errorMsg += ` ${passwordCheck.feedback.warning}.`;
+      }
+      if (passwordCheck.feedback.suggestions.length > 0) {
```
i feel like feedback suggestions for a password, esp displayed as a toast, doesn't send the right ux message. this is more of a master password than an auth password, so we should just have the user retry

AndrewCLu commented:
> <img alt="Screenshot 2024-11-03 at 9 06 16 PM" width="429" src="https://private-user-images.githubusercontent.com/7270633/382691992-243d6c2c-e648-40c1-bf17-40322357cda7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3MTQyMjYsIm5iZiI6MTczMDcxMzkyNiwicGF0aCI6Ii83MjcwNjMzLzM4MjY5MTk5Mi0yNDNkNmMyYy1lNjQ4LTQwYzEtYmYxNy00MDMyMjM1N2NkYTcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMDRUMDk1MjA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Yzk1YjYxYzFjNThlMWM1Y2RmZDUyYmYzYWNhNDY0ZjY1NzA4NWM3MDYzZDk4YTZhNGEwMTJiOTU3NjAzZjczMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.HZBF9TDartEhsMvuqcHetuKAtATbH8VkMOBjkMKEaDo"> <img alt="Screenshot 2024-11-03 at 9 06 03 PM" width="432" src="https://private-user-images.githubusercontent.com/7270633/382692012-04fa5967-da3c-46b7-be7b-4c6835769121.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3MTQyMjYsIm5iZiI6MTczMDcxMzkyNiwicGF0aCI6Ii83MjcwNjMzLzM4MjY5MjAxMi0wNGZhNTk2Ny1kYTNjLTQ2YjctYmU3Yi00YzY4MzU3NjkxMjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMDRUMDk1MjA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDI4ZWQyODRkMjRlMWRhZWNhNTkyY2YyZjkyMGJjZGI0NzVmODdlNzAwMWM2MTNiZDhkYmU0ZTFkZjM5ZTdiYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.xE6UIcltsOV-lv5q1g2FcAqihKTbfaAaytwzy6BTD0o">

This message is kinda jank. Can we instead just have a generic toast that says something like "use less common words, add numbers and symbols, etc."

stevenelleman commented:
> This message is kinda jank. Can we instead just have a generic toast that says something like "use less common words, add numbers and symbols, etc."

I don't think you pulled the most recent change. I fixed the styling: https://github.com/cursive-team/connections/pull/165#issuecomment-2454202250

Updated to remove suggestion, so it's not that relevant anyways: 
<img width="423" alt="image" src="https://github.com/user-attachments/assets/4d18654a-ae61-49be-80e6-55382a536902">

AndrewCLu reviewed (approved):
Update toast text to "Weak password, try adding numbers, symbols, and using less common words."
